hbase-server/src/main/java/org/apache/hadoop/hbase/wal/AbstractWALRoller.java

```diff
@@ -249,4 +244,42 @@ public void close() {
     running = false;
     interrupt();
   }
+
+  /**
+   * Independently control the roll of each wal. When use multiwal,
+   * can avoid all wal roll together. see HBASE-24665 for detail
+   */
+  protected class RollController {
+    private final WAL wal;
+    private boolean isRequestRoll;
+    private long lastRollTime;
+
+    RollController(WAL wal) {
+      this.wal = wal;
+      this.isRequestRoll = false;
+      this.lastRollTime = System.currentTimeMillis();
+    }
+
+    public synchronized void requestRoll() {
+      this.isRequestRoll = true;
+    }
+
+    public synchronized Map<byte[], List<byte[]>> rollWal(long lastRollTime) throws IOException {
```

Comment:
This is overall coming good now.  Much more clean than what we had before. Thanks
One issue here though.  We have rollWal() and requestRoll() under same lock now.. That is not good. The requestRoll should not wait for ongoing roll to be finished. The request call is in hot path.